api/src/soc_mgr.rs

```diff
@@ -60,4 +203,155 @@ pub trait SocManager {
+            if timeout_cycles == 0 {
+                return Err(CaliptraApiError::MailboxTimeout);
+            }
+        }
+        let status = self.soc_mbox().status().read().status();
+        if status.cmd_failure() {
+            self.soc_mbox().execute().write(|w| w.execute(false));
+            let soc_ifc = self.soc_ifc();
+            return Err(CaliptraApiError::MailboxCmdFailed(
+                if soc_ifc.cptra_fw_error_fatal().read() != 0 {
+                    soc_ifc.cptra_fw_error_fatal().read()
+                } else {
+                    soc_ifc.cptra_fw_error_non_fatal().read()
+                },
+            ));
+        }
+        if status.cmd_complete() {
+            self.soc_mbox().execute().write(|w| w.execute(false));
+            return Ok(None);
+        }
+        if !status.data_ready() {
+            return Err(CaliptraApiError::UnknownCommandStatus(status as u32));
+        }
+
+        let res = mbox_read_response(self.soc_mbox(), resp_data);
+
+        self.soc_mbox().execute().write(|w| w.execute(false));
+
+        let buf = res?;
+
+        Ok(Some(buf))
+    }
+
+    /// Executes a typed request and (if success), returns the typed response.
+    /// The checksum field of the request is calculated, and the checksum of the
+    /// response is validated.
+    fn mailbox_exec_req<R: Request>(
+        &mut self,
+        mut req: R,
+        resp_bytes: &mut [u8],
+    ) -> core::result::Result<R::Resp, CaliptraApiError> {
+        if mem::size_of::<R>() < mem::size_of::<MailboxReqHeader>() {
+            return Err(CaliptraApiError::MailboxReqTypeTooSmall);
+        }
+        if mem::size_of::<R::Resp>() < mem::size_of::<MailboxRespHeader>() {
+            return Err(CaliptraApiError::MailboxRespTypeTooSmall);
+        }
+        if R::Resp::MIN_SIZE < mem::size_of::<MailboxRespHeader>() {
+            return Err(CaliptraApiError::MailboxRespTypeTooSmall);
+        }
+        let (header_bytes, payload_bytes) = req
+            .as_bytes_mut()
+            .split_at_mut(mem::size_of::<MailboxReqHeader>());
+
+        let mut header = MailboxReqHeader::read_from(header_bytes as &[u8]).unwrap();
+        header.chksum = calc_checksum(R::ID.into(), payload_bytes);
+        header_bytes.copy_from_slice(header.as_bytes());
+
+        let Some(data) = SocManager::mailbox_exec(self, R::ID.into(), req.as_bytes(), resp_bytes)? else {
+                return Err(CaliptraApiError::MailboxNoResponseData);
+        };
+
+        if data.len() < R::Resp::MIN_SIZE || data.len() > mem::size_of::<R::Resp>() {
+            return Err(CaliptraApiError::MailboxUnexpectedResponseLen {
+                expected_min: R::Resp::MIN_SIZE as u32,
+                expected_max: mem::size_of::<R::Resp>() as u32,
+                actual: data.len() as u32,
+            });
+        }
+
+        let mut response = R::Resp::new_zeroed();
+        response.as_bytes_mut()[..data.len()].copy_from_slice(data);
```

Comment:
Is it possible to also use `read_from_prefix` here?
```suggestion
        let response = R::Resp::read_from_prefix(data).unwrap();
```

Reply:
This actually causes panic when the response is variable size. I changed HwModel to delegate the mailbox_execute_request to SocMgr and it panicked in this line when processing caliptra_api::mailbox::GetLdevCertResp.